src/Orleans/IDs/SiloAddress.cs

```diff
@@ -253,7 +253,48 @@ public bool Equals(SiloAddress other)
 
         public int CompareTo(SiloAddress other)
         {
-            return other == null ? 1 : Generation.CompareTo(other.Generation);
+            if (other == null) return 1;
+            // Compare Generation first. It gives a cheap and fast way to compare, avoiding allocations 
+            // and is also semantically meaningfull - older silos (with smaller Generation) will appear first in the comparison order.
+            // Only if Generations are the same, go on to compare Ports and IPAddress (which is more expansive to compare).
+            // Alternatively, we could compare ConsistentHashCode or UniformHashCode.
+            int comp = Generation.CompareTo(other.Generation);
+            if (comp != 0) return comp;
+
+            comp = Endpoint.Port.CompareTo(other.Endpoint.Port);
+            if (comp != 0) return comp;
+
+            return CompareIpAddresses(Endpoint.Address, other.Endpoint.Address);
+        }
+
+        // Taken from: http://www.codeproject.com/Articles/26550/Extending-the-IPAddress-object-to-allow-relative-c
+        private static int CompareIpAddresses(IPAddress one, IPAddress two)
+        {
+            int returnVal = 0;
+            if (one.AddressFamily == two.AddressFamily)
+            {
+                byte[] b1 = one.GetAddressBytes();
+                byte[] b2 = two.GetAddressBytes();
```

Comment:
hmmm, good point, if as @gabikliot mentions, this path is not executed extremely often, we could always map it to IPv6 (as we can't just do it when the address does not match, because of the rules of transitivity in how `CompareTo` is supposed to work).
We could even have a Lazy<byte[]> field with the result of doing `ipAddress.MapToIPv6().GetAddressBytes()` for the cases when it does run often.

Maybe @sebastianburckhardt has feedback on how often this comparison will be used as opposed to just returning after comparing `Generation`